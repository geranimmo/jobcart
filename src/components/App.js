import React, { Component } from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import reducers from './reducers';
import RouterComponent from './Routers';

export class App extends Component {
	render() {
		const store = createStore(reducers, {}, applyMiddleware(thunk));

		return (
			<Provider store={store}>
				<RouterComponent />
			</Provider>
		);
	}
}

export default App;