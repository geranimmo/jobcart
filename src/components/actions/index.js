export * from './Authentication';
export * from './GetListPackage';
export * from './AddToCart';
export * from './RemoveFromCart';
export * from './GetTotalCost';
export * from './GetTotalDiscount';
