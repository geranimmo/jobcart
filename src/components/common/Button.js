import React from "react";

const Button = props => {
	return <button type="button" {...props}>{props.value}</button>;
};

export { Button };
