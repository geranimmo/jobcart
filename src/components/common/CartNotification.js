import React from "react";

const CartNotification = props => {
	const {
		totalItems,
		className
	} = props;
	const cartTotal = totalItems > 9 ? '9+' : totalItems;
    
	return <div className={`${className} ${totalItems === 0 ? 'hidden' : ''}`}>{cartTotal}</div>;
};

export { CartNotification };
