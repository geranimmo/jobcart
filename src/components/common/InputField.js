import React from "react";

const InputField = props => {
	return <input {...props}/>;
};

export { InputField };
