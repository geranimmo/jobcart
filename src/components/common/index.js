export { InputField } from './InputField';
export { Button } from './Button';
export { CartFooter } from './CartFooter';
export { DateTimeConverter } from './DateTimeConverter';
export { SubstringText } from './SubstringText';
export { CartItem } from './CartItem';
export { CartNotification } from './CartNotification';