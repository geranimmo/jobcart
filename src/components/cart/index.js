import React, { Component } from 'react';
import { withRouter } from "react-router-dom";
import { connect } from 'react-redux';
import _ from 'lodash';
import Header from '../header';
import { CartItem, CartFooter } from '../common';
import { removeFromCart } from '../actions';
import './cart.less';

export class Cart extends Component {
	constructor(props) {
		super(props);
		this.state = {
			myCartList: [],
			totalCartDiscount: 0,
			totalCartCost: 0
		};
	}
    
	componentDidMount() {
		if (!this.props.UserLogin) {
			this.props.history.push('/');
		}
		
		document.addEventListener('scroll', this.handleScroll, true);
	}
    
	componentWillMount() {
		if (this.props.UserLogin) {
			const { ShoppingCart, Packages } = this.props;
			const myCartArr = _.map(ShoppingCart, (item) => {
				return _.merge(item, _.find(Packages, { 'package_id' : item.id }));
			});

			this.setState({ myCartList: myCartArr });

			setTimeout(() => {
				this.populatePromoItems();
				this.populateTotalCost();
			}, 0);
		}
	}
    
    handleScroll = () => {
    	let targetElm = document.getElementById("content__scroller").getBoundingClientRect();
    	let targetOffsetTop = targetElm.top;
    	let header = document.getElementById("header");
		
    	if (targetOffsetTop < 0) {
    		header.classList.add("header__on__scrolled");
    	} else {
    		header.classList.remove("header__on__scrolled");
    	}
    }
	
	removeCartItems = (data) => {
		if (this.props.UserLogin) {
			this.props.removeFromCart(data);

			const updateCartList = this.state.myCartList.filter(e => e.buy_time !== data);

			this.setState({ myCartList: updateCartList });
			
			setTimeout(() => {
				this.populatePromoItems();
				this.populateTotalCost();
			}, 0);
		}
	};

	populateTotalCost() {
		const { myCartList } = this.state;
		let totalCost = 0;
	
		if ( myCartList.length > 0 ) {
			for ( let i = 0; i < myCartList.length; i++ ) {
				totalCost = parseFloat((totalCost + myCartList[i].package_price).toFixed(2));
			}
		}
	
		this.setState({ totalCartCost: totalCost });
	}

	populatePromoItems = () => {
		const { myCartList } = this.state;
		const { client_special } = this.props.Profile;
		let totalDiscount = 0;
	
		if ( client_special.length > 0 ) {
			for ( let i = 0; i < client_special.length; i++ ) {
				let getInCartPromoItems = [];
				
				if ( client_special[i].types === 'free__items' ) {
					// CALCULATE FREE-FOR-SAME-AD
					getInCartPromoItems = filterPromoById(client_special[i].package_id, myCartList);
	
					if (getInCartPromoItems.length > 0) {
						let pricePerItem = getInCartPromoItems[0].package_price;
						let getTotalFreeItems = Math.floor(filterPromoById(client_special[i].package_id, myCartList).length / client_special[i].minimum_items);
						totalDiscount = parseFloat((getTotalFreeItems * pricePerItem).toFixed(2));
					}
				} else if ( client_special[i].types === 'discount__items' ) {
					// CALCULATE DISC-FOR-MIN-BUY
					getInCartPromoItems = filterPromoById(client_special[i].package_id, myCartList);
					
					if (getInCartPromoItems.length >= client_special[i].minimum_items) {
						let discountPrice = client_special[i].discount_price;
						totalDiscount = parseFloat((discountPrice * getInCartPromoItems.length).toFixed(2));
					}
				}
			}
		}
	
		this.setState({ totalCartDiscount: totalDiscount });
	};

	render() {
    	const { myCartList } = this.state;

    	return (
    		<div className={`main__container`}>
    			<Header onCart={true} />
    			<main id={this.props.id} className={`cart__component`}>
    				<div
    					id={`content__scroller`}
    					onScroll={this.handleScroll.bind(this)}
    				>
    					<div className={`cart__empty ${myCartList.length ? 'hidden' : ''}`}>Your Cart Still Empty</div>;
    					{ myCartList.map((item, index) => {
    						return (
								<div key={index}>
									<CartItem
										id={`Cart__${item.id}__${index}__${item.buy_time}`}
										className={`cart__list__wrapper`}
										removeCartItems={this.removeCartItems.bind(this, item.buy_time)}
										item={item}
									/>
								</div>
							);
    					}) }
    				</div>
    			</main>
    			<CartFooter className={`cart__footer`} {...this.state}/>
    		</div>
    	);
	}
}

export const filterPromoById = (id, datas) => {
	return datas.filter(obj => obj.package_id === id);
};

const mapStateToProps = state => {
	return state;
};

export default withRouter(connect(mapStateToProps, { removeFromCart })(Cart));