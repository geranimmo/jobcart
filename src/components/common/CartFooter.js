import React from 'react';
import { Button } from './Button';

const CartFooter = props => {
	let fixedCost = parseFloat((props.totalCartCost - props.totalCartDiscount).toFixed(2));
	
	const buttonProps = {
		className: 'button__checkout',
		value: `Checkout (${props.myCartList.length})`
	};
	
	return (
		<div className={props.className}>
			<div>
				<span>
					<label>Total Cost:</label>
					${fixedCost}
				</span>
				<Button {...buttonProps}/>
			</div>
		</div>
	);
};

export { CartFooter };