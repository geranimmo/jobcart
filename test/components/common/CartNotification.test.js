import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { CartNotification } from '../../../src/components/common';

describe('>>> C A R T  N O T I F I C A T I O N ---- Test & Snapshot <<<', () => {
	let wrapper;
	const propsValue = {
		className: `cart__notif`,
		totalItems: 10
	};

	beforeEach(() => {
		wrapper = shallow(<CartNotification {...propsValue}/>);
	});
       
	it('+++ Contains props that assigned and display total text 9+ +++', () => {
		const targetComponent = wrapper.find(`div.${propsValue.className}`).prop('className');

		expect(targetComponent).toContain(`${propsValue.className}`);
		expect(wrapper.find(`div.${propsValue.className}`).text()).toBe('9+');
	});
    
	it('+++ Should contain hidden in className props if total is 0 +++', () => {
		const sampleValue = {
			className: `cart__notif`,
			totalItems: 0
		};
		const sampleWrapper = shallow(<CartNotification {...sampleValue}/>);
		const targetComponent = sampleWrapper.find(`div.${propsValue.className}`).prop('className');

		expect(targetComponent).toContain('hidden');
	});

	it('+++ Capturing Snapshot of CartNotification +++', () => {
		const renderedValue = renderer.create(<CartNotification {...propsValue}/>).toJSON();
		
		expect(renderedValue).toMatchSnapshot();
	});
});