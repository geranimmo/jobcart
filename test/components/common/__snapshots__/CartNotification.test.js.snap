// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> C A R T  N O T I F I C A T I O N ---- Test & Snapshot <<< +++ Capturing Snapshot of CartNotification +++ 1`] = `
<div
  className="cart__notif "
>
  9+
</div>
`;
